oap-server/server-receiver-plugin/envoy-metrics-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/envoy/als/K8SServiceRegistry.java

```diff
@@ -0,0 +1,217 @@
+                params.resourceVersion,
+                params.timeoutSeconds,
+                params.watch,
+                null
+            ),
+            V1Endpoints.class,
+            V1EndpointsList.class
+        ).addEventHandler(new ResourceEventHandler<V1Endpoints>() {
+            @Override
+            public void onAdd(final V1Endpoints endpoints) {
+                addEndpoints(endpoints);
+            }
+
+            @Override
+            public void onUpdate(final V1Endpoints oldEndpoints, final V1Endpoints newEndpoints) {
+                addEndpoints(newEndpoints);
+            }
+
+            @Override
+            public void onDelete(final V1Endpoints endpoints, final boolean deletedFinalStateUnknown) {
+                removeEndpoints(endpoints);
+            }
+        });
+    }
+
+    private void listenPodEvents(final CoreV1Api coreV1Api, final SharedInformerFactory factory) {
+        factory.sharedIndexInformerFor(
+            params -> coreV1Api.listPodForAllNamespacesCall(
+                null,
+                null,
+                null,
+                null,
+                null,
+                null,
+                params.resourceVersion,
+                params.timeoutSeconds,
+                params.watch,
+                null
+            ),
+            V1Pod.class,
+            V1PodList.class
+        ).addEventHandler(new ResourceEventHandler<V1Pod>() {
+            @Override
+            public void onAdd(final V1Pod pod) {
+                addPod(pod);
+            }
+
+            @Override
+            public void onUpdate(final V1Pod oldPod, final V1Pod newPod) {
+                addPod(newPod);
+            }
+
+            @Override
+            public void onDelete(final V1Pod pod, final boolean deletedFinalStateUnknown) {
+                removePod(pod);
+            }
+        });
+    }
+
+    private void removePod(final V1Pod pod) {
+        log.debug("Removing pod {}", pod);
+
+        Optional.ofNullable(pod.getStatus()).ifPresent(
+            status -> ipServiceMap.remove(status.getPodIP())
+        );
+    }
+
+    private void addPod(final V1Pod pod) {
+        log.debug("Adding pod {}", pod);
+
+        Optional.ofNullable(pod.getStatus()).ifPresent(
+            status -> {
+                final String ip = status.getPodIP();
+                final ServiceMetaInfo service = ipServiceMap.computeIfAbsent(ip, unused -> new ServiceMetaInfo());
+
+                final V1ObjectMeta podMeta = requireNonNull(pod.getMetadata());
+                service.setServiceInstanceName(String.format("%s.%s", podMeta.getName(), podMeta.getNamespace()));
+                service.setTags(transformLabelsToTags(podMeta.getLabels()));
+            }
+        );
+    }
+
+    private void addEndpoints(final V1Endpoints endpoints) {
+        log.debug("Adding endpoints {}", endpoints);
+
+        final String serviceName = requireNonNull(endpoints.getMetadata()).getName();
```

Comment:
We had better introduce a service name rule to format it. For instance, if some users want to append a version to the service name, they have to issue an expression, like `${service.name}-${pod.labels.version}`

If they deploy `bookinfo`, the service name of the service productpage might be `productpage-v1`, `producatpage-v2` and etc.

Reply:
I think the key is providing the format expression. Defaut should be only service name with version, correct?

Reply:
Done. I’ll add some more tests of this later. Please review again @hanahmily , thanks